### PR TITLE
[6.1] Workflow emails

### PR DESCRIPTION
**Pull request by roland-d**
Pull Request for Issue #37745 #37446 .

### Summary of Changes
This change adds the option to select email as a notification type. Users can send notifications to:
- Private messages (old and default behavior)
- Email

image

### Testing Instructions
First you need to apply this PR. Since there are database changes you will need to run those as well. 

For MySQL users run this on your database:
```
INSERT IGNORE INTO `#__mail_templates` (`template_id`, `extension`, `language`, `subject`, `body`, `htmlbody`, `attachments`, `params`) VALUES
('plg_workflow_notification.mail', 'plg_workflow_notification', '', 'PLG_WORKFLOW_NOTIFICATION_EMAIL_ON_TRANSITION_SUBJECT', 'PLG_WORKFLOW_NOTIFICATION_EMAIL_ON_TRANSITION_MSG', '', '', '{"tags":["siteurl","title","performer","transitionName","toStage","extraText"]}');
```

For Postgres users run this on your database:
```
INSERT INTO "#__mail_templates" ("template_id", "extension", "language", "subject", "body", "htmlbody", "attachments", "params") VALUES
('plg_workflow_notification.mail', 'plg_workflow_notification', '', 'PLG_WORKFLOW_NOTIFICATION_EMAIL_ON_TRANSITION_SUBJECT', 'PLG_WORKFLOW_NOTIFICATION_EMAIL_ON_TRANSITION_MSG', '', '', '{"tags":["siteurl","title","performer","transitionName","toStage","extraText"]}')
ON CONFLICT DO NOTHING;
```

1. Enable Workflows by going to Content -> Articles -> Options -> Integration -> Enable Workflow and set it to Yes
2. Save the changes
3. Go to Content -> Workflows
4. Click on the number 7 (or another number) under Transitions
5. Click on the Unpublish transition
6. Click on Notification
7. Notice there is now the option to select a notification type
8. The `Private Messages` is the old and default way of sending notifications. Clicking on the dropdown now also let's you select `Email` as an option. Multiple options are possible as well
9. Select a usergroup and/or a user where you want to send the message to
10. Save the transition
11. Go got Content -> Articles
12. Edit an article
13. Set the transition to `Unpublished`
14. Save the article
15. Check if the message has arrived

#### Mail templates
The sending of the emails uses the mail templates in Joomla. The mail template can be customized via these steps:
1. Go to System -> Mail Templates
2. Filter on `workflow`
3. You will see 1 template. 
image
4. Click on the template to edit it
5. Modify the template to your needs

### Feedback requested
The following questions I have:
1. The name of the person doing the transition has a placeholder named `performer` Any suggestions for a better name?
2. Are there more placeholders wanted for the mail template?

@brianteeman Could you please check the language files?

#### Final words
Thanks to @bembelimen and @softforge for the idea to add this. Thanks to @brianteeman to point out to use mail templates.

I know this isn't perfect code so looking for feedback to get this into Joomla to make transitions more useful as they are quite neat.

### Actual result BEFORE applying this Pull Request
Messages are only sent to users who can receive private messages

### Expected result AFTER applying this Pull Request
Messages can now also be send to users with an account and email address.

### Documentation Changes Required
Yes, need to document the addition of the email option.

### CODE REVIEW

**Comment by Kostelano** [comment]
Great addition.
I tested a little, I'll leave a comment.

The "Additional Message Text" field must be removed. First, we find ourselves in a similar situation, which I described in the issue https://github.com/joomla/joomla-cms/issues/37493. Secondly, if we add text in this field and we select both methods of receiving a notification, then we will receive this additional text in personal messages, but we will NOT receive it in email, since this additional text will not be in the letter template. Thirdly, in this field we do not see the real message (only the description), but we see it in the email template.

In short, we really need to leave 1 place for setting the "body" of the notification and this place in the email templates. We can mention this in the description for the "Notification Type" parameter.

**Comment by brianteeman** [comment]
Doesn't make any sense to keep the old broken behaviour as an option You're introducing a new method that fixes the bugs in the old method but still leave the old method there with its bugs

File: plugins/workflow/notification/forms/action.xml [hunk 1]
```diff
@@ -13,6 +13,22 @@
...
+					name="notification_type"
+					type="list"
+					label="PLG_WORKFLOW_NOTIFICATION_TYPE_LABEL"
+					description="PLG_WORKFLOW_NOTIFICATION_TYPE_DESC"
+					layout="joomla.form.field.list-fancy-select"
+					showon="notification_send_mail:1"
+					multiple="true"
+					default="messages"
+					required="true"
+			>
```
Comment (brianteeman) [inline plugins/workflow/notification/forms/action.xml]:
```suggestion
				name="notification_type"
				type="list"
				label="PLG_WORKFLOW_NOTIFICATION_TYPE_LABEL"
				description="PLG_WORKFLOW_NOTIFICATION_TYPE_DESC"
				layout="joomla.form.field.list-fancy-select"
				showon="notification_send_mail:1"
				multiple="true"
				default="messages"
				required="true"
				>
```

**Comment by brianteeman** [comment]
You cannot have a required field that is conditionaly displayed. 

You have made the notification type a required field but is only shown when send email is set to yes.

**Comment by brianteeman** [comment]
I really dont undrstand why you have kept the broken private messages version. There is no explanation or why you would use method email or method private messags and you can still select groups and users to receive private messages which will neveer be sent.

**Comment by brianteeman** [comment]
It needs to be clearer that you can customise the mail further if you are using the email method by using th email templates BUT that it is a generic template for all transition notification emails and you must put anything specific to the transition in the notification config only.

**Comment by brianteeman** [comment]
In what scenario would anyone ever send a message by email AND private message?

File: administrator/language/en-GB/plg_workflow_notification.ini [hunk 2]
```diff
@@ -9,12 +9,20 @@ PLG_WORKFLOW_NOTIFICATION="Workflow - Notification"
 PLG_WORKFLOW_NOTIFICATION_ADDTEXT="The stage has changed"
 PLG_WORKFLOW_NOTIFICATION_ADDTEXT_DESC="This text will be sent: Title [title], changed by [user], new state: [state]. You can add your own text to this message and you can localise the text by using a language key and creating a language override."
 PLG_WORKFLOW_NOTIFICATION_ADDTEXT_LABEL="Additional Message Text"
+PLG_WORKFLOW_NOTIFICATION_EMAIL_ON_TRANSITION_MSG="Title: {TITLE}. Transition {TRANSITIONNAME} performed by {PERFORMER}. New state: {TOSTAGE}."
```
Comment (brianteeman) [inline administrator/language/en-GB/plg_workflow_notification.ini]: USER not PERFORMER
A transition does not = a change in stage

Comment (roland-d) [reply administrator/language/en-GB/plg_workflow_notification.ini]: I have changed PERFORMER to USER. As for the rest of the message, this is a verbatim copy of what is send to a private message. Any changes to the text I can include of course.

**Comment by roland-d** [comment]
@Kostelano 
> The "Additional Message Text" field must be removed.

This cannot be removed as it will be a backwards compatible break.

> Secondly, if we add text in this field and we select both methods of receiving a notification, then we will receive this additional text in personal messages, but we will NOT receive it in email, since this additional text will not be in the letter template. 

You can add the additional text to the email. This is a valid point that the extra text should be in the default mail template. I have updated this PR with that change.

> Thirdly, in this field we do not see the real message (only the description), but we see it in the email template.

I do not understand this. Can you explain this further?

> In short, we really need to leave 1 place for setting the "body" of the notification and this place in the email templates.

That is something I was thinking about but really have no idea how we should do this. First I cannot remove this addition message text due to backwards compatibility break. Second, we could create a mail template for every transition but these cannot be in core because we have no idea which transitions there will be.

@brianteeman 

> Doesn't make any sense to keep the old broken behaviour as an option 

I cannot remove sending messages to private messages because that would be a backwards compatible break. I would not say it is broken but rather limited as we can only send messages to users with access to private messages.

> You cannot have a required field that is conditionaly displayed.

Actually you can. The required condition is only active when the field is displayed.

> There is no explanation or why you would use method email or method private messags 

Any suggestions for a text to add?

> you can still select groups and users to receive private messages which will neveer be sent.

I saw your other issue about that but this is not in scope of this PR. I can look at that in a separate PR.

> It needs to be clearer that you can customise the mail further if you are using the email method by using th email templates BUT that it is a generic template for all transition notification emails and you must put anything specific to the transition in the notification config only.

See my remark above for Kostelano, he basically says the same thing. Any suggestions on how to improve this are welcome.

> In what scenario would anyone ever send a message by email AND private message?

I knew this was coming :) If I chose a single dropdown the request would come to make it multiple, make it multiple and the request will come for a single dropdown. Ghehehehe.

So my reasoning is that the transition message can be send but may not arrive in the mail but you can see it in your private message box. Why limit the options for the user? We can let the user decide what they want to use, we do not have to decide for them.

**Comment by brianteeman** [comment]
> > Doesn't make any sense to keep the old broken behaviour as an option

> I cannot remove sending messages to private messages because that would be a backwards compatible break. I would not say it is broken but rather limited as we can only send messages to users with access to private messages.

No its not  B/C break you are just changing from a completely broken method to a working method

>> You cannot have a required field that is conditionaly displayed.

> Actually you can. The required condition is only active when the field is displayed.

I do not write things just for the sake of writing words. I tested the PR. (and I know that we already have the same bug elsewhere.)

image

image

>> you can still select groups and users to receive private messages which will neveer be sent.

> I saw your other issue about that but this is not in scope of this PR. I can look at that in a separate PR.

Of course it is in the scope of this PR. Otherwise it doesn't fix the bug you state this pr fixes.

**Comment by brianteeman** [comment]
@Kostelano you need the "additional text" field. Without it you cannot have a unique message for each notification. 

Mail templates are for general customisations of all notification emails

**Comment by roland-d** [comment]
> No its not B/C break you are just changing from a completely broken method to a working method

It works on the site where we are using workflow. Messages go to the private message box.

> I do not write things just for the sake of writing words. I tested the PR. (and I know that we already have the same bug elsewhere.)

I also tested this of course but did not read between your lines that there should not be an option selected. I removed the requirement.

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on a94c82124dd67a8eb622e708783c1403a872bd4a<br /><br />Does not fix the reported problems <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37860">issues.joomla.org/tracker/joomla-cms/37860</a>.</sub>

File: administrator/language/en-GB/plg_workflow_notification.ini [hunk 3]
```diff
@@ -9,12 +9,20 @@ PLG_WORKFLOW_NOTIFICATION="Workflow - Notification"
 PLG_WORKFLOW_NOTIFICATION_ADDTEXT="The stage has changed"
 PLG_WORKFLOW_NOTIFICATION_ADDTEXT_DESC="This text will be sent: Title [title], changed by [user], new state: [state]. You can add your own text to this message and you can localise the text by using a language key and creating a language override."
 PLG_WORKFLOW_NOTIFICATION_ADDTEXT_LABEL="Additional Message Text"
+PLG_WORKFLOW_NOTIFICATION_EMAIL_ON_TRANSITION_MSG="Title: {TITLE}. Transition {TRANSITIONNAME} performed by {USER}. New state: {TOSTAGE}.{EXTRATEXT}"
```
Comment (Kostelano) [inline administrator/language/en-GB/plg_workflow_notification.ini]:
```suggestion
PLG_WORKFLOW_NOTIFICATION_EMAIL_ON_TRANSITION_MSG="Title: {TITLE}. Transition {TRANSITIONNAME} performed by {USER}. New state: {TOSTAGE}. {EXTRATEXT}"
```

**Comment by Kostelano** [comment]
If you add some text in the transition field, then it will be sent to the mail with a visible br tag, although I just added the text in the field. This is the EMAIL method.

Screenshot_1

There is no br tag in private messages.

File: plugins/workflow/notification/forms/action.xml [hunk 4]
```diff
@@ -13,6 +13,21 @@
...
+				type="list"
+				label="PLG_WORKFLOW_NOTIFICATION_TYPE_LABEL"
+				description="PLG_WORKFLOW_NOTIFICATION_TYPE_DESC"
+				layout="joomla.form.field.list-fancy-select"
+				showon="notification_send_mail:1"
+				multiple="true"
+				default="messages"
+				>
+					<option value="messages">PLG_WORKFLOW_NOTIFICATION_TYPE_MESSAGES</option>
+					<option value="email">PLG_WORKFLOW_NOTIFICATION_TYPE_EMAIL</option>
```
Comment (brianteeman) [inline plugins/workflow/notification/forms/action.xml]:
```suggestion
				<option value="messages">PLG_WORKFLOW_NOTIFICATION_TYPE_MESSAGES</option>
				<option value="email">PLG_WORKFLOW_NOTIFICATION_TYPE_EMAIL</option>
```

**Comment by brianteeman** [comment]
How about updating the text below the selection of the notification type with something like

image

It wont prevent someone trying to send to users who won't receive the private messages email but at least they have been warned

File: administrator/language/en-GB/plg_workflow_notification.ini [hunk 5]
```diff
@@ -9,12 +9,20 @@ PLG_WORKFLOW_NOTIFICATION="Workflow - Notification"
...
+PLG_WORKFLOW_NOTIFICATION_MAIL_MAIL_DESC="Mail sent to the selected usergroup and/or users set for the transition"
+PLG_WORKFLOW_NOTIFICATION_MAIL_MAIL_TITLE="Workflow: Transition Mail"
+PLG_WORKFLOW_NOTIFICATION_NO_TITLE="Unknown title"
 PLG_WORKFLOW_NOTIFICATION_ON_TRANSITION_MSG="Title: %1$s. Transition \"%2$s\" performed by %3$s. New state: %4$s."
 PLG_WORKFLOW_NOTIFICATION_ON_TRANSITION_SUBJECT="The status of \"%s\" has been changed"
-PLG_WORKFLOW_NOTIFICATION_NO_TITLE="Unknown title"
 PLG_WORKFLOW_NOTIFICATION_RECEIVERS_LABEL="Users"
 PLG_WORKFLOW_NOTIFICATION_SENDMAIL_LABEL="Send Notification"
 PLG_WORKFLOW_NOTIFICATION_SENT="Notifications sent"
+PLG_WORKFLOW_NOTIFICATION_TYPE_DESC="Select one or more notification types you would like to use"
```
Comment (Quy) [inline administrator/language/en-GB/plg_workflow_notification.ini]: Maybe remove as it provides no additional information.

Comment (Quy) [inline administrator/language/en-GB/plg_workflow_notification.ini]: Maybe remove as it provides no additional information.

Comment (roland-d) [reply administrator/language/en-GB/plg_workflow_notification.ini]: I have updated it with the text provided by Brian above.

**Comment by roland-d** [comment]
> There is no br tag in private messages.

Good catch. I moved it to the private message where it comes from.

So I have updated this PR to also check if there is any recipient selected when saving the transition and if not, an error message is shown and the transition is not saved. The same goes if for some reason a user manages to save with no notification types selected. That should fix the issue of enabling the notification but not selecting any recipients.

**Comment by richard67** [comment]
@roland-d I've allowed myself to fix the conflict in the supports.sql files.

**Comment by roland-d** [comment]
Rebased to 4.3

**Comment by spamzini** [comment]
Hi,

i'v upgraded a CLEAN installation of joomla 4.2.0 to the last one 4.2.2.
I'v upgrade by the file upload (administrator/index.php?option=com_joomlaupdate&view=upload) and not by the automatic update
The check of the database by the internal function .../administrator/index.php?option=com_installer&view=database shows two errors of database: 
One of this was the XXXX_mail_templates has MEDIUMTEXT character set.
I solved with this SQL command:

ALTER TABLE `XXXX_mail_templates` CHANGE `htmlbody` `htmlbody` MEDIUMTEXT CHARACTER SET utf8 COLLATE utf8_unicode_ci NOT NULL;<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37860">issues.joomla.org/tracker/joomla-cms/37860</a>.</sub>

**Comment by richard67** [comment]
@roland-d As the PR has been rebased to 4.3-dev, the update SQL scripts "4.2.0-2022-05-22.sql" need to be renamed to **something later than** the up to now latest 4.3 update SQL scripts "4.3.0-2022-09-23.sql". I would suggest "4.3.0-2022-10-23.sql".

**Comment by richard67** [comment]
@obuisard Please don't merge this before the update SQL scripts have been renamed, see my previous comment.

**Comment by obuisard** [comment]
> @obuisard Please don't merge this before the update SQL scripts have been renamed, see my previous comment.

No worries Richard, I was not going to merge it as I saw your comments :-)
Thank you for your help!

**Comment by Magnytu2** [comment]
I have tested this item :red_circle: unsuccessfully on cd97f4b1a0e7c39abc0120a2ed6da88e3fd76cb9<br /><br />Still no mail received after a transition. The private message notification works, but not the email.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37860">issues.joomla.org/tracker/joomla-cms/37860</a>.</sub>

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.0-dev. No new features will be merged into Joomla! 4.3 series. Joomla! 4.4 series is a bridge release to make migration from Joomla! 4 to 5 as smooth as possible.

**Comment by micker** [comment]
hello i would help to test this but after save my item i have this message "No notifications sent as there are no users to send this message to." bug or need to configure something ?
image
regards

**Comment by micker** [comment]
I have tested this item :white_check_mark: successfully on cd97f4b1a0e7c39abc0120a2ed6da88e3fd76cb9<br /><br />thanks @Richardnet its great now
@roland-d tested and validate !

maybe 2 return :smile: 

- how to customize email peer transition (for now one mail for all)
- direct link to item can be more usefull (not only to admin)
thanks for all<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37860">issues.joomla.org/tracker/joomla-cms/37860</a>.</sub>

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.1-dev.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.2-dev.

**Comment by carcam** [comment]
I'm trying to apply the patch to test this but I get this error:

> The file marked for modification does not exist: plugins/workflow/notification/notification.php

It seems the notification plugin was updated to the new architecture and you need to update your pr @roland-d . Can you please check?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37860">issues.joomla.org/tracker/joomla-cms/37860</a>.</sub>

**Comment by roland-d** [comment]
Just a heads up, I have merged the 5.2 dev into this branch but I have not tested anything yet. More to follow later.

File: plugins/workflow/notification/src/Extension/Notification.php [hunk 6]
```diff
@@ -19,10 +19,13 @@
 use Joomla\CMS\Language\Text;
 use Joomla\CMS\Mail\MailTemplate;
 use Joomla\CMS\Plugin\CMSPlugin;
+use Joomla\CMS\User\UserFactoryAwareTrait;
```
Comment (brianteeman) [inline plugins/workflow/notification/src/Extension/Notification.php]: please sort alphabetically

File: plugins/workflow/notification/src/Extension/Notification.php [hunk 7]
```diff
@@ -19,10 +19,13 @@
...
 use Joomla\CMS\Uri\Uri;
 use Joomla\CMS\User\UserFactoryInterface;
 use Joomla\CMS\Workflow\WorkflowPluginTrait;
 use Joomla\CMS\Workflow\WorkflowServiceInterface;
+use Joomla\Database\DatabaseAwareTrait;
+use Joomla\Event\DispatcherInterface;
```
Comment (brianteeman) [inline plugins/workflow/notification/src/Extension/Notification.php]: please sort the use statements alphabetically

**Comment by cybersalt** [comment]
I have tested this item :red_circle: unsuccessfully on 1f172e818f874c3f6431b22a666abde5eec6138a<br /><br />I followed all of the steps - selecting the notification to go to the superuser group.

Upon saving the article this was the error message:

Menu
Frontend
An error has occurred.
0 Joomla\Plugin\Workflow\Notification\Extension\Notification::getUsersFromGroup(): Argument #1 ($data) must be of type Joomla\Plugin\Workflow\Notification\Extension\stdClass, stdClass given, called in /home/basicjoo/j5core.basicjoomla.com/stageit/plugins/workflow/notification/src/Extension/Notification.php on line 195
Call Stack
#	Function	Location
1	()	JROOT/plugins/workflow/notification/src/Extension/Notification.php:392<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/37860">issues.joomla.org/tracker/joomla-cms/37860</a>.</sub>

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 5.3-dev.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 6.0-dev.

**Comment by HLeithner** [comment]
This pull request has been automatically rebased to 6.1-dev.
